### ISSUE TITLE
Executing the command “.\CreateTestCertificate.ps1 -AddAsTrustedRootAuthority -UseLocalMachineStore” of “CreateTestCertificate.ps1” in Entropy repo failed at encryption on Windows

### ISSUE THREAD

**Issue by v-luzh** [NuGet/Home#12969]
### NuGet Product Used

Other/NA

### Product Version

latest downloaded Entropy repo

### Impact

None

#### Repro Steps:
1.	Open “Task Manager” on a VM with OS Windows-11-Enterprise. 
2.	Click the button “Run new task”  on the top-right of the window.
3.	In the Open textbox type:       
       powershell -ExecutionPolicy Bypass 
4.	Enable the option “Create this task with administrative privileges.” 
5.	Click OK. 
6.	Cd to the path of the VM , run: Git clone https://github.com/NuGet/Entropy.git and run below command.
7.	Cd to <path_to_Entropy>\TestCertGenerator and run `“.\CreateTestCertificate.ps1 -AddAsTrustedRootAuthority -UseLocalMachineStore”`.

#### Expected:
The command can be executed correctly without any error and the certificate is created successfully into the path of “Entropy\TestCertGenerator”.
#### Actual:
Executing the command “.\CreateTestCertificate.ps1 -AddAsTrustedRootAuthority -UseLocalMachineStore” of “CreateTestCertificate.ps1” in Entropy repo failed at encryption as below screenshot and there is no certificate created under the path of “Entropy\TestCertGenerator”.
![image](https://github.com/NuGet/Home/assets/20632625/2711ebee-d498-44e2-8991-d817422373c8)
#### Note:
1.It reproes on 80% Windows VMs.
2.It didn't repro on the Entropy repo downloaded on Sept.20th.

**Comment by kartheekp-ms** [comment]
I am able to reproduce this issue.

> It didn't repro on the Entropy repo downloaded on Sept.20th.

Looking at the commit history, it looks like there were no new commits after Sep 8, 2023.

**Comment by nkolev92** [comment]
@v-luzh Does reruning the script fix the problem? Is this blocking your tests?

**Comment by nkolev92** [comment]
@kartheekp-ms If this is blocking our tests, then we need to get a fix soon. 
Can you please take a look to see if this is something that's easily fixable?

**Comment by v-luzh** [comment]
> @v-luzh Does reruning the script fix the problem? Is this blocking your tests?

Hi @nkolev92, re-running the script would not work. 

For the testing on Oct.26th, it reproduced on 80% of the VMs (it doesn't repro on only one VM), so we went on executing the rest test steps of that case on the working VM. It may block the case next time if we don't have a VM which the script ".\CreateTestCertificate.ps1 -AddAsTrustedRootAuthority -UseLocalMachineStore" is OK to run on. So we think it need a quick fix.

**Comment by jeffkl** [comment]
Team Triage: @kartheekp-ms could you please take a look and see if its something simple that we can fix?

**Comment by v-luzh** [comment]
Hi @nkolev92 @kartheekp-ms, we found it will also repro when running the command `.\CreateTestCertificate.ps1 -AddAsTrustedRootAuthority` on Windows. And that will **block** our quarterly **Package Signing Test**. Please help to give a quick fix on this issue. Thanks a lot.

**Comment by kartheekp-ms** [comment]
@v-luzh I am able to reproduce this issue almost every time on my local machine. Unfortunately, I don't have a fix right now, but I think we have a workaround, which is to pass the `-Password` option to the script. Does this workaround work for you?

```
~\source\repos\Entropy\TestCertGenerator❯ .\CreateTestCertificate.ps1 -UseLocalMachineStore -AddAsTrustedRootAuthority -Password abc
Creating a new test certificate for testing NuGet package signing.
Certificate created with fingerprint:  abcdefghi12345
Certificate added to certificate store:  LocalMachine\Root
Certificate exported to:  .\abcdefghi12345.pfx
```

@dtivel - II think we need your help here. The [Export-PfxCertificate](https://learn.microsoft.com/powershell/module/pki/export-pfxcertificate?view=windowsserver2022-ps) command is failing with an `Encryption Failed.` exception (see the image in the issue description) when the `-ProtectTo $currentUser` parameter is used. This issue was reported [previously](https://github.com/NuGet/Home/issues/9798#issuecomment-660182931) but it did not hinder manual testing. However, now this issue is reproducible almost every time and may block testing in the near future, as @v-luzh mentioned in the previous [comment](https://github.com/NuGet/Home/issues/12969#issuecomment-1831357684). As I mentioned earlier, using the `Password` parameter works.

I'm not sure if it matters, but according to the documentation, the [NotBefore](https://learn.microsoft.com/powershell/module/pki/new-selfsignedcertificate?view=windowsserver2022-ps#-notbefore) property is set to 10 minutes before the certificate creation time, whereas in [our script, we use DateTime.Now](https://github.com/NuGet/Entropy/blob/main/TestCertGenerator/CreateTestCertificate.ps1#L4). I changed this value `DateTime.Now.AddMinutes(-10)` to match the default value in my testing but didn't make any difference. I would appreciate any suggestions you might have.

**Comment by dtivel** [comment]
Why are we using CreateTestCertificate.ps1 anymore?  I'm reluctant to spend any cycles on it.  Do we have instructions somewhere that still mention CreateTestCertificate.ps1?

Use https://github.com/NuGet/Entropy/tree/main/MakeTestCert to create test certificates and https://github.com/NuGet/Entropy/tree/main/TrustTestCert to add/remove trust.

**Comment by v-luzh** [comment]
Hi @dtivel, we still use "CreateTestCertificate.ps1" because the [Cross-platform Dotnet NuGet Sign Test](https://microsoft.sharepoint.com/teams/NuGet/_layouts/15/Doc.aspx?sourcedoc={8a3c6685-1661-47f4-8807-b765ca62aea1}&action=edit&wd=target%28Package%20Signing%20Manual%20Test.one%7C0344a15b-01b1-4828-af4c-63fe53687f8a%2FCross-platform%20Dotnet%20NuGet%20Sign%20Test%7C6ab81998-1393-4354-9783-c76fbc90277e%2F%29&wdorigin=703) and [Package Signing Test](https://microsoft.sharepoint.com/teams/NuGet/_layouts/15/Doc.aspx?sourcedoc={8a3c6685-1661-47f4-8807-b765ca62aea1}&action=edit&wd=target%28Package%20Signing%20Manual%20Test.one%7C0344a15b-01b1-4828-af4c-63fe53687f8a%2FPackage%20Signing%20Test%7Cebfadab8-c4b3-4b12-bf73-ab8e56a7dfd8%2F%29&wdorigin=703) depend on it.

And we investigated to use the https://github.com/NuGet/Entropy/tree/main/MakeTestCert to create test certificates, only this case is failed as below. 
![image](https://github.com/NuGet/Home/assets/20632625/025a9944-25fb-4668-bd9f-d9a294fd1434)
Could you please help to verify how to provide a option by using "[MakeTestCert](https://github.com/NuGet/Entropy/tree/main/MakeTestCert)" for these cases:[ Dotnet nuget sign command:  sign when certificate is not valid for code signing](https://microsoft.sharepoint.com/teams/NuGet/_layouts/15/Doc.aspx?sourcedoc={8a3c6685-1661-47f4-8807-b765ca62aea1}&action=edit&wd=target%28Package%20Signing%20Manual%20Test.one%7C0344a15b-01b1-4828-af4c-63fe53687f8a%2FDotnet%20nuget%20sign%20command%20%20sign%20when%20certificate%20is%20not%20%7Cb9343605-98df-4d9b-8a66-42cf468c29fc%2F%29&wdorigin=703) & [sign command:  sign when certificate is not valid for code signing](https://microsoft.sharepoint.com/teams/NuGet/_layouts/15/Doc.aspx?sourcedoc={8a3c6685-1661-47f4-8807-b765ca62aea1}&action=edit&wd=target%28Package%20Signing%20Manual%20Test.one%7C0344a15b-01b1-4828-af4c-63fe53687f8a%2F536701%20sign%20command%20%20sign%20when%20certificate%20is%20not%20valid%20%7Cb435c767-ee35-46b4-b4cd-6435a1d7eaed%2F%29&wdorigin=703)? Thanks.

**Comment by v-luzh** [comment]
> @v-luzh I am able to reproduce this issue almost every time on my local machine. Unfortunately, I don't have a fix right now, but I think we have a workaround, which is to pass the `-Password` option to the script. Does this workaround work for you?
> 
> ```
> ~\source\repos\Entropy\TestCertGenerator❯ .\CreateTestCertificate.ps1 -UseLocalMachineStore -AddAsTrustedRootAuthority -Password abc
> Creating a new test certificate for testing NuGet package signing.
> Certificate created with fingerprint:  abcdefghi12345
> Certificate added to certificate store:  LocalMachine\Root
> Certificate exported to:  .\abcdefghi12345.pfx
> ```
> 
> @dtivel - II think we need your help here. The [Export-PfxCertificate](https://learn.microsoft.com/powershell/module/pki/export-pfxcertificate?view=windowsserver2022-ps) command is failing with an `Encryption Failed.` exception (see the image in the issue description) when the `-ProtectTo $currentUser` parameter is used. This issue was reported [previously](https://github.com/NuGet/Home/issues/9798#issuecomment-660182931) but it did not hinder manual testing. However, now this issue is reproducible almost every time and may block testing in the near future, as @v-luzh mentioned in the previous [comment](https://github.com/NuGet/Home/issues/12969#issuecomment-1831357684). As I mentioned earlier, using the `Password` parameter works.
> 
> I'm not sure if it matters, but according to the documentation, the [NotBefore](https://learn.microsoft.com/powershell/module/pki/new-selfsignedcertificate?view=windowsserver2022-ps#-notbefore) property is set to 10 minutes before the certificate creation time, whereas in [our script, we use DateTime.Now](https://github.com/NuGet/Entropy/blob/main/TestCertGenerator/CreateTestCertificate.ps1#L4). I changed this value `DateTime.Now.AddMinutes(-10)` to match the default value in my testing but didn't make any difference. I would appreciate any suggestions you might have.

Hi @kartheekp-ms, thanks for your investigation. The workaround works for us. Just want to confirm with you all, should we use the workaround or use the https://github.com/NuGet/Entropy/tree/main/MakeTestCert to create test certificates (has an issue as above) in the following [Cross-platform Dotnet NuGet Sign Test](https://microsoft.sharepoint.com/teams/NuGet/_layouts/15/Doc.aspx?sourcedoc=%7B8a3c6685-1661-47f4-8807-b765ca62aea1%7D&action=edit&wd=target%28Package%20Signing%20Manual%20Test.one%7C0344a15b-01b1-4828-af4c-63fe53687f8a%2FCross-platform%20Dotnet%20NuGet%20Sign%20Test%7C6ab81998-1393-4354-9783-c76fbc90277e%2F%29&wdorigin=703) and [Package Signing Test](https://microsoft.sharepoint.com/teams/NuGet/_layouts/15/Doc.aspx?sourcedoc=%7B8a3c6685-1661-47f4-8807-b765ca62aea1%7D&action=edit&wd=target%28Package%20Signing%20Manual%20Test.one%7C0344a15b-01b1-4828-af4c-63fe53687f8a%2FPackage%20Signing%20Test%7Cebfadab8-c4b3-4b12-bf73-ab8e56a7dfd8%2F%29&wdorigin=703) tests?

**Comment by kartheekp-ms** [comment]
> Hi @kartheekp-ms, thanks for your investigation. The workaround works for us. 

Happy to hear that the workaround works and can unblock manual testing efforts in the future, in case the test script starts to fail on all the VMs.

> Just want to confirm with you all, should we use the workaround or use the https://github.com/NuGet/Entropy/tree/main/MakeTestCert to create test certificates (has an issue as above)

Damon is the SME on package signing functionality. Therefore, I suggest following his guidance, which means updating the test steps to use the https://github.com/NuGet/Entropy/blob/main/MakeTestCert tool for creating test certificates. The [README.md](https://github.com/NuGet/Entropy/blob/main/MakeTestCert/README.md) clearly explains the several options currently supported by the new tool. In your comment, you mentioned that `has an issue as above`. Are you indicating that the new tool is also failing to create and export self-signed test certificates? Until we are 100% certain that all steps work properly with the new tool, we can continue relying on the PowerShell scripts currently used for our day-to-day testing. Please let us know if you need our help in updating or reviewing the new test steps. 

cc @v-luzh

**Comment by v-luzh** [comment]
Hi @kartheekp-ms, now there is only one option -Type `SSLServerAuthentication` (as I highlighted with red line in the screenshot above) which does not supported with the new tool. So, if we use the new tool (update the test cases now), these two cases:[ Dotnet nuget sign command: sign when certificate is not valid for code signing](https://microsoft.sharepoint.com/teams/NuGet/_layouts/15/Doc.aspx?sourcedoc=%7B8a3c6685-1661-47f4-8807-b765ca62aea1%7D&action=edit&wd=target%28Package%20Signing%20Manual%20Test.one%7C0344a15b-01b1-4828-af4c-63fe53687f8a%2FDotnet%20nuget%20sign%20command%20%20sign%20when%20certificate%20is%20not%20%7Cb9343605-98df-4d9b-8a66-42cf468c29fc%2F%29&wdorigin=703) & [sign command: sign when certificate is not valid for code signing](https://microsoft.sharepoint.com/teams/NuGet/_layouts/15/Doc.aspx?sourcedoc=%7B8a3c6685-1661-47f4-8807-b765ca62aea1%7D&action=edit&wd=target%28Package%20Signing%20Manual%20Test.one%7C0344a15b-01b1-4828-af4c-63fe53687f8a%2F536701%20sign%20command%20%20sign%20when%20certificate%20is%20not%20valid%20%7Cb435c767-ee35-46b4-b4cd-6435a1d7eaed%2F%29&wdorigin=703) will be blocked.

**Comment by nkolev92** [comment]
Do we still want to keep those 2 tests @dtivel? 

Sounds like we might need some additions to the tool.

**Comment by dtivel** [comment]
I updated [MakeTestCert](https://github.com/NuGet/Entropy/tree/main/MakeTestCert) to support creation of TLS certificates.  An example is given in the updated help output (as well as in the README.md):

```
  MakeTestCert.exe -eku 1.3.6.1.5.5.7.3.1 -eku 1.3.6.1.5.5.7.3.2
    Creates an RSA 3072-bit TLS certificate valid for 2 hours from creation time
```

Note that MakeTestCert is a .NET app which requires the .NET SDK to build.  You can use any .NET 7 or 8 SDK.  However, to ensure that the .NET SDK you use isn't accidentally the .NET SDK that you test (unless that's what you want), I recommend installing the .NET SDK to a local directory (vs. a global installation).  You can do this using [`dotnet-install.ps1`'s `-InstallDir` option](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script).

I did not update test case steps to use MakeTestCert instead of the CreateTestCertificate.ps1 script.  That would take me a lot of time, which I currently don't have.  Any one of you (@v-luzh, @kartheekp-ms, etc.) can do this.  Or, I will find time at a later date to do it.  I don't think you're blocked from running these tests if you use [MakeTestCert](https://github.com/NuGet/Entropy/tree/main/MakeTestCert) and [TrustTestCert](https://github.com/NuGet/Entropy/tree/main/TrustTestCert).

> Do we still want to keep those 2 tests @dtivel?

Sure, why not.

@v-luzh, if you're unblocked, please close this issue.  Otherwise, please let me know how I can help unblock you.

**Comment by v-luzh** [comment]
@dtivel, thanks for fixing this tool and it works now for the "-eku" option as below screenshot.
![image](https://github.com/NuGet/Home/assets/20632625/02242158-107b-4562-b848-dd545face797)
I'm closing this bug as "fixed" now. We will update the test cases by ourselves and get back to you if we meet any issues.
